feat: add C `ndarray` implementation for `stats/base/dmeanvarpn`

DhruvArvindSingh opened the pull request:
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds C `ndarray` implementation for `stats/base/dmeanvarpn`.
-   Updates README.md.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DhruvArvindSingh committed:
feat: add C ndarray implementation for stats/base/dmeanvarpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/benchmark/c/benchmark.length.c:
```diff
@@ -94,9 +94,12 @@ static double rand_double( void ) {
+static double benchmark1( int iterations, int len ) {
 	double elapsed;
 	double out[ 2 ];
+		if(len <= 0){
```
removing `if ` results in cpplint error for uninitialized `x`.

aayush0325 replied:
`// cppcheck-suppress uninitvar`

use this

DhruvArvindSingh replied:
Thanks for the help

DhruvArvindSingh commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/benchmark/c/benchmark.length.c:
```diff
@@ -122,6 +125,44 @@ static double benchmark( int iterations, int len ) {
+static double benchmark2( int iterations, int len ) {
+	double elapsed;
+	double out[ 2 ];
+	if(len <= 0){
```
removing `if ` results in cpplint error for uninitialized `x`.

aayush0325 replied:
`// cppcheck-suppress uninitvar`

use this

aayush0325 replied:
see [stats/base/dmax/benchmark/c/benchmark.length.c](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/base/dmax/benchmark/c/benchmark.length.c)

DhruvArvindSingh replied:
Thanks for the help

DhruvArvindSingh commented:
/stdlib update-copyright-years

DhruvArvindSingh committed:
fix: updated readme.md and benchmark.length.c

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

DhruvArvindSingh committed:
Merge branch 'stdlib-js:develop' into feat/ndarray_dmenvarpn

DhruvArvindSingh committed:
refactor: refactored stats/base/dmeanvarpn

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/README.md:
````diff
@@ -239,6 +239,115 @@ console.log( out );
+Computes the [mean][arithmetic-mean] and [variance][variance] of a double-precision floating-point strided array using a two-pass algorithm.
+
+```c
+double x[] ={ 1.0, -2.0, 2.0 };
+double out[2];
+stdlib_strided_dmeanvarpn( x.length, 1, x, 1, out, 1 );
+// Out = [ ~0.3333, ~4.3333 ]
````
```suggestion
const double x[] ={ 1.0, -2.0, 2.0 };
double out[] = { 0.0, 0.0 };

stdlib_strided_dmeanvarpn( x.length, 1, x, 1, out, 1 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/README.md:
```diff
@@ -239,6 +239,115 @@ console.log( out );
+-   **correction**: `[in] double` degrees of freedom adjustment.
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **Out**: `[in] double*` Output array.
```
```suggestion
-   **Out**: `[out] double*` Output array.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/README.md:
````diff
@@ -239,6 +239,115 @@ console.log( out );
+Computes the [mean][arithmetic-mean] and [variance][variance] of a double-precision floating-point strided array using a two-pass algorithm and alternative indexing semantics.
+
+```c
+const double x[] = { 1.0, -2.0, 2.0 };
+
+stdlib_strided_dmeanvarpn_ndarray( x.length, 1, x, 1, 0, out, 1, 0 );
+// Out = [ ~0.3333, ~4.3333 ]
````
no declaration for `Out` here

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/README.md:
```diff
@@ -239,6 +239,115 @@ console.log( out );
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **offsetX**: `[in] CBLAS_INT` `X` starting index.
+-   **Out**: `[in] double*` Output array.
```
```suggestion
-   **Out**: `[out] double*` Output array.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/README.md:
this is also missing changes in the JavaScript docs regarding removing the usage of `math/base/special/floor` in examples

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/benchmark/benchmark.js:
```diff
@@ -39,15 +45,8 @@ var dmeanvarpn = require( './../lib/dmeanvarpn.js' );
-		x[ i ] = ( randu()*20.0 ) - 10.0;
-	}
-	out = new Float64Array( 2 );
+	var out = uniform( 2, 0.0, 0.0, options );
```
```suggestion
	var out = new Float64Array( 2 );
```

aayush0325 replied:
same comment for all other JS benchmarks

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/docs/repl.txt:
```diff
@@ -32,13 +32,13 @@
 
     strideX: integer
-        Index increment for `x`.
+        Stride Lenght for `x`.
```
```suggestion
        Stride length for `x`.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/docs/repl.txt:
```diff
@@ -32,13 +32,13 @@
 
     strideOut: integer
-        Index increment for `out`.
+        Stride Lenght for `out`.
```
```suggestion
        Stride length for `out`.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/docs/repl.txt:
```diff
@@ -97,7 +96,7 @@
 
     strideX: integer
-        Index increment for `x`.
+        Stride Length for `x`.
```
```suggestion
        Stride length for `x`.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/docs/repl.txt:
```diff
@@ -106,7 +105,7 @@
 
     strideOut: integer
-        Index increment for `out`.
+        Stride Length for `out`.
```
```suggestion
        Stride length for `out`.
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/examples/c/example.c:
```diff
@@ -17,22 +17,21 @@
 
 	// Specify the number of elements:
-	int64_t N = 4;
+	int N = 4;
```
```suggestion
	const int N = 4;
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/examples/c/example.c:
```diff
@@ -17,22 +17,21 @@
 	// Specify the stride lengths:
-	int64_t strideX = 2;
-	int64_t strideOut = 1;
+	int strideX = 2;
+	int strideOut = 1;
```
```suggestion
	const int strideX = 2;
	const int strideOut = 1;
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/examples/index.js:
```diff
@@ -18,21 +18,16 @@
 console.log( x );
 
-out = new Float64Array( 2 );
+var out = discreteUniform( 10, 0, 0, {
+	'dtype': 'float64'
+});
```
```suggestion
var out = new Float64Array( 2 );
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/lib/dmeanvarpn.native.js:
```diff
@@ -43,7 +43,7 @@ var addon = require( './../src/addon.node' );
 *
 * var v = dmeanvarpn( x.length, 1, x, 1, out, 1 );
-* // returns <Float64Array>[ ~0.3333, ~4.3333 ]
+*  //returns <Float64Array>[ ~0.3333, ~4.3333 ]
```
```suggestion
* // returns <Float64Array>[ ~0.3333, ~4.3333 ]
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/manifest.json:
this is an incorrect format for `manifest.json` for this type of implementation, please see an existing implementation to get an idea of what we're looking for

aayush0325 replied:
https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/stats/base/dmin/manifest.json

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/src/addon.c:
```diff
@@ -17,7 +17,6 @@
 */
 
 #include "stdlib/stats/base/dmeanvarpn.h"
-#include "stdlib/blas/base/shared.h"
```
why is this being removed? it should be there since we're wrapping function calls in `API_SUFFIX` which comes from this header

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/src/main.c:
```diff
@@ -16,72 +16,76 @@
 */
 
-#include "stdlib/stats/base/dmeanvarpn.h"
+
```
```suggestion
```
no need for extra newline

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/src/main.c:
```diff
@@ -16,72 +16,76 @@
 /**
 * Computes the mean and variance of a double-precision floating-point strided array using a two-pass algorithm.
 *
-* ## Method
-*
-* -   This implementation uses a two-pass approach, as suggested by Neely (1966).
-*
-* ## References
-*
-* -   Neely, Peter M. 1966. "Comparison of Several Algorithms for Computation of Means, Standard Deviations and Correlation Coefficients." _Communications of the ACM_ 9 (7). Association for Computing Machinery: 496–99. doi:[10.1145/365719.365958](https://doi.org/10.1145/365719.365958).
-* -   Schubert, Erich, and Michael Gertz. 2018. "Numerically Stable Parallel Computation of (Co-)Variance." In _Proceedings of the 30th International Conference on Scientific and Statistical Database Management_. New York, NY, USA: Association for Computing Machinery. doi:[10.1145/3221269.3223036](https://doi.org/10.1145/3221269.3223036).
-*
```
references are important, no need to remove them

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/src/main.c:
```diff
@@ -16,72 +16,76 @@
-	} else {
-		ix = 0;
-	}
-	if ( strideOut < 0 ) {
-		io = -strideOut;
-	} else {
-		io = 0;
-	}
+	ix = offsetX;
+	io = offsetOut;
 	if ( N <= 0 ) {
-		Out[ io ] = 0.0 / 0.0; // NaN
-		Out[ io+strideOut ] = 0.0 / 0.0; // NaN
+		Out[ io ] = 0.0/0.0;
+		Out[ io + strideOut ] = 0.0/0.0;
 		return;
 	}
-	dN = (double)N;
-	n = dN - correction;
+	n = N - correction;
 	if ( N == 1 || strideX == 0 ) {
 		Out[ io ] = X[ ix ];
 		if ( n <= 0.0 ) {
-			Out[ io+strideOut ] = 0.0 / 0.0; // NaN
+			Out[ io+strideOut ] = 0.0/0.0;
 		} else {
 			Out[ io+strideOut ] = 0.0;
 		}
```
this is copied from the JS ndarray implementation, i would prefer that you refactor the existing C implementation only

aayush0325 replied:
this would avoid implicit typecasting etc, for eg see changes on line 101

aayush0325 commented on lib/node_modules/@stdlib/stats/base/dmeanvarpn/src/main.c:
```diff
@@ -16,72 +16,76 @@
 		Out[ io ] = X[ ix ];
 		if ( n <= 0.0 ) {
-			Out[ io+strideOut ] = 0.0 / 0.0; // NaN
+			Out[ io+strideOut ] = 0.0/0.0;
```
```suggestion
			Out[ io+strideOut ] = 0.0 / 0.0; // NaN
```

aayush0325 replied:
same comment for all other instances
